internal/wasm/interpreter/interpreter.go

```diff
@@ -188,57 +170,58 @@ type interpreterOp struct {
+		irs, err := wazeroir.CompileFunctions(e.enabledFeatures, module)
+		if err != nil {
+			return err
+		}
+		for i, ir := range irs {
+			compiled, err := e.lowerIR(ir)
+			if err != nil {
+				return fmt.Errorf("function[%d/%d] failed to convert wazeroir operations: %w", i, len(module.FunctionSection)-1, err)
+			}
+			funcs = append(funcs, compiled)
+		}
+	}
+	e.addCompiledFunctions(module, funcs)
+	return nil
+
+}
+
 // NewModuleEngine implements the same method as documented on wasm.Engine.
-func (e *engine) NewModuleEngine(name string, source *wasm.Module, importedFunctions, moduleFunctions []*wasm.FunctionInstance, table *wasm.TableInstance, tableInit map[wasm.Index]wasm.Index) (wasm.ModuleEngine, error) {
+func (e *engine) NewModuleEngine(name string, module *wasm.Module, importedFunctions, moduleFunctions []*wasm.FunctionInstance, table *wasm.TableInstance, tableInit map[wasm.Index]wasm.Index) (wasm.ModuleEngine, error) {
 	imported := uint32(len(importedFunctions))
 	me := &moduleEngine{
 		name:                  name,
-		compiledFunctions:     make([]*compiledFunction, 0, imported+uint32(len(moduleFunctions))),
 		parentEngine:          e,
 		importedFunctionCount: imported,
 	}
 
-	for idx, f := range importedFunctions {
-		cf, ok := e.getCompiledFunction(f)
-		if !ok {
-			return nil, fmt.Errorf("import[%d] func[%s]: uncompiled", idx, f.DebugName)
-		}
+	for _, f := range importedFunctions {
+		cf := f.Module.Engine.(*moduleEngine).compiledFunctions[f.Index]
 		me.compiledFunctions = append(me.compiledFunctions, cf)
 	}
 
-	if cached, ok := e.getCachedCompiledFunctions(source); ok { // cache hit
-		for i, c := range cached[len(importedFunctions):] {
-			cloned := c.clone(me, moduleFunctions[i])
-			me.compiledFunctions = append(me.compiledFunctions, cloned)
-		}
-	} else { // cache miss
-		for i, f := range moduleFunctions {
-			var compiled *compiledFunction
-			if f.Kind == wasm.FunctionKindWasm {
-				ir, err := wazeroir.Compile(e.enabledFeatures, f)
-				if err != nil {
-					me.Close()
-					// TODO(Adrian): extract Module.funcDesc so that errors here have more context
-					return nil, fmt.Errorf("function[%d/%d] failed to lower to wazeroir: %w", i, len(moduleFunctions)-1, err)
-				}
-
-				compiled, err = e.lowerIROps(f, ir.Operations)
-				if err != nil {
-					me.Close()
-					return nil, fmt.Errorf("function[%d/%d] failed to convert wazeroir operations: %w", i, len(moduleFunctions)-1, err)
-				}
-			} else {
-				compiled = &compiledFunction{hostFn: f.GoFunc, source: f}
-			}
-			compiled.moduleEngine = me
-			me.compiledFunctions = append(me.compiledFunctions, compiled)
-
-			// Add the compiled function to the store-wide engine as well so that
-			// the future importing module can refer the function instance.
-			e.addCompiledFunction(f, compiled)
-		}
+	compiledFunctions, ok := e.getCompiledFunctions(module)
+	if !ok {
+		return nil, fmt.Errorf("source module for %s must be compiled before instantiation", name)
```

Comment:
I think we don't need to introduce source since it is the same type still. Maybe since this is a bug on our side, say "BUG: didn't call CompileModule on module %s", name

Reply:
well it is actually not a bug on ourside as some one might accidentally try to call `InstantiateModule` _after_ calling `CompiledCode.Close`